Fix manylinux wheel builds

sklam opened the pull request:
llvmdev:

- bump to llvm15
- drop zstd due to conda package uses too new of cxxabi. If needed, we'll need to build it for manylinux ourselves.

docker image:
- x86-64 uses manylinux2014 (which is glibc2.17)
- aarch64 uses manylinux_2_28 since conda doesn't have glibc2.17 builds

misc:
- renamed old directory that assumes manylinux_2_17

sklam commented:
BFID `llvmlite_wheel_70`

esc commented on buildscripts/manylinux/docker_run.sh:
```diff
@@ -0,0 +1,11 @@
+
+echo "MINICONDA_FILE=$MINICONDA_FILE"
+# Ensure the latest docker image
+image="quay.io/pypa/${MANYLINUX_IMAGE}:latest"
```
variables in shell scripts should be uppercase by convention, so in this case I would recommend: `$IMAGE`

esc replied:
or maybe even `$IMAGE_URI`?

esc commented on conda-recipes/llvmdev_manylinux/build.sh:
```diff
@@ -0,0 +1,116 @@
+
+set -x
+
+# Make osx work like linux.
```
More details about what this does? It should be evident from the comment what the purpose of this line is.

esc replied:
Since we are building only on linux here, can these lines not just be deleted?

esc commented on conda-recipes/llvmdev_manylinux/build.sh:
```diff
@@ -0,0 +1,116 @@
+fi
+
+# disable -fno-plt due to https://bugs.llvm.org/show_bug.cgi?id=51863 due to some GCC bug
+if [[ "$target_platform" == "linux-ppc64le" ]]; then
```
Can this be removed or commented out, since we don't support `linux-ppc64le`?

esc commented on conda-recipes/llvmdev_manylinux/build.sh:
```diff
@@ -0,0 +1,116 @@
+  CMAKE_ARGS="${CMAKE_ARGS} -DFFI_LIBRARY_DIR=$PREFIX/lib"
+fi
+
+if [[ $target_platform == osx-arm64 ]]; then
```
Can this be removed or commented out, since we don't support `osx-arm64`?

esc commented on conda-recipes/llvmdev_manylinux/build.sh:
```diff
@@ -0,0 +1,116 @@
+
+ninja install
+
+if [[ "${target_platform}" == "linux-64" || "${target_platform}" == "osx-64" ]]; then
```
Can this be removed or commented out, since we don't support osx-64?

esc commented on conda-recipes/llvmdev_manylinux/build.sh:
```diff
@@ -0,0 +1,116 @@
+'ExecutionEngine/Interpreter/intrinsics.ll'
+  fi
+
+  if [[ "$target_platform" == osx-* ]]; then
```
Can this be removed or commented out, since we don't support osx-*?

esc commented on conda-recipes/llvmdev_manylinux/meta.yaml:
```diff
@@ -0,0 +1,74 @@
+    - ../llvm15-remove-use-of-clonefile.patch
+    - ../llvm15-svml.patch
+    - ../compiler-rt-cfi-startproc-war.patch
+    - ../compiler-rt-macos-build.patch
```
This looks like a `macos` patch? Since this is manylinux -- can that line be removed?

sklam replied:
This meta.yaml file is a copy of `../llvmdev/meta.yaml`. I will comment out lines that already have preprocessor conditions for non-linux. But, for patches, i will want to keep them the same as the non-manylinux version for ease of maintenance (so the code is the same in build variants)

esc commented on conda-recipes/llvmdev_manylinux/meta.yaml:
```diff
@@ -0,0 +1,74 @@
+    - python >=3
+    # - libcxx      # it is not defined{{ cxx_compiler_version }}  # [osx]
+    - patch       # [not win]
+    - m2-patch    # [win]
```
This seems windows specific, can this line be removed? Or commented out?

esc commented on conda-recipes/llvmdev_manylinux/meta.yaml:
```diff
@@ -0,0 +1,74 @@
+    #- libcxx    # it is not defined{{ cxx_compiler_version }}  # [osx]
+    - libffi    # [unix]
+    # libxml2 supports a windows-only feature, see https://github.com/llvm/llvm-project/blob/llvmorg-17.0.6/llvm/include/llvm/WindowsManifest/WindowsManifestMerger.h
+    - libxml2   # [win]
```
This seems windows specific, can this line be removed? Or commented out?

esc commented on conda-recipes/llvmdev_manylinux/meta.yaml:
```diff
@@ -0,0 +1,74 @@
+    - $PREFIX/bin/llvm-config --libs                         # [not win]
+    - $PREFIX/bin/llc -version                               # [not win]
+
+    - if not exist %LIBRARY_INC%\\llvm\\Pass.h exit 1        # [win]
```
This seems windows specific, can this line be removed? Or commented out?

esc commented on conda-recipes/llvmdev_manylinux/meta.yaml:
```diff
@@ -0,0 +1,74 @@
+    - $PREFIX/bin/llc -version                               # [not win]
+
+    - if not exist %LIBRARY_INC%\\llvm\\Pass.h exit 1        # [win]
+    - if not exist %LIBRARY_LIB%\\LLVMSupport.lib exit 1     # [win]
```
This seems windows specific, can this line be removed? Or commented out?

esc commented on conda-recipes/llvmdev_manylinux/meta.yaml:
```diff
@@ -0,0 +1,74 @@
+
+    # LLD tests
+    - ld.lld --version                                       # [unix]
+    - lld-link /?                                            # [win]
```
This seems windows specific, can this line be removed? Or commented out?

esc reviewed (changes requested):
I've given this a first pass, there are numerous questions and suggestions,mostly related to code cleanliness, overall coherence and adherence to conventions. One they are resolved I will do a functional review and actually test the code to ensure it behaves as advertised.

esc commented on conda-recipes/llvmdev_manylinux/build.sh:
```diff
@@ -0,0 +1,116 @@
+#!/bin/bash
+
+# based on https://github.com/AnacondaRecipes/llvmdev-feedstock/blob/master/recipe/build.sh
```
PS: Maybe include a link to the commit hash or version here?

sklam replied:
This file is a copy of `../llvmdev/build.sh` with minor changes. I'll add file level comments to document this.

sklam commented on buildscripts/manylinux/README.md:
```diff
@@ -1,29 +1,30 @@
-- 64-bit linux: `./buildscripts/manylinux_2_17/docker_run_x64.sh build_llvmdev.sh`
-- aarch64 linux: `./buildscripts/manylinux_2_17/docker_run_aaarch64.sh build_llvmdev.sh`
+- x86_64 linux: `./buildscripts/manylinux/docker_run_x64.sh build_llvmdev.sh`
+    - uses manylinux2014 image for glibc 2.17+
```
`manylinux2014` is renamed to `manylinux` for "future-proof"  and that `aarch64` is not `manylinux2014` so the directories are renamed to avoid confusion.

sklam commented on buildscripts/manylinux/build_llvmlite.sh:
```diff
@@ -17,22 +17,24 @@ outputdir="/root/llvmlite/docker_output"
+pip install setuptools
+
 # Clean up
-git clean -xdf llvmlite build
```
git is not happy with the docker mapped directory and can fail the build sometimes. Also, the `setup.py clean` later does all of the needed cleaning.

sklam commented on buildscripts/manylinux/docker_run.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+set -xe
+# Use this to make the llvmdev packages that are manylinux2010 compatible
```
it's outdated comment.

sklam commented on buildscripts/manylinux/docker_run_aarch64.sh:
```diff
@@ -1,5 +1,4 @@
-export PRECMD="linux32"
-export MINICONDA_FILE="https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86.sh"
+export MANYLINUX_IMAGE="manylinux_2_28_aarch64"
+export MINICONDA_FILE="https://repo.anaconda.com/miniconda/Miniconda3-py311_24.9.2-0-Linux-aarch64.sh"
```
no particular reason. Just the first link i found probably because 311 is sorted first. It shouldn't matter.

sklam commented on buildscripts/manylinux/build_llvmlite.sh:
```diff
@@ -17,22 +17,24 @@ outputdir="/root/llvmlite/docker_output"
+    conda install -y numba/label/manylinux_2_28::llvmdev --no-deps
 else
-    conda install -y -c numba/label/manylinux2014 llvmdev
+    conda install -y numba/label/manylinux_2_17::llvmdev --no-deps
```
changed in https://github.com/numba/llvmlite/pull/1099/commits/7ba4e73e2ce5fe1a36238ab5a7452f7944adaa0f

sklam commented on buildscripts/manylinux/build_llvmdev.sh:
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+set -xe
+cd $(dirname $0)
+source ./prepare_miniconda.sh $1
```
Added error checking in https://github.com/numba/llvmlite/pull/1099/commits/7ba4e73e2ce5fe1a36238ab5a7452f7944adaa0f

sklam commented:
Addressed all review comments. The scripts have room for improvements but we should just do them later as part of updating all the CI/release infrastructure.

esc commented:
> Addressed all review comments. The scripts have room for improvements but we should just do them later as part of updating all the CI/release infrastructure.

I looked at all the changes you made and am happy with the response.

esc reviewed (approved):
I spot checked this using docker on mac.

First I ran:

```
./buildscripts/manylinux/docker_run_x64.sh build_llvmdev.sh
```

This produced the `llvmdev` package locally.

Then I applied the patch:

```
diff --git i/buildscripts/manylinux/build_llvmlite.sh w/buildscripts/manylinux/build_llvmlite.sh
index 582b35c991..73f1c35313 100755
--- i/buildscripts/manylinux/build_llvmlite.sh
+++ w/buildscripts/manylinux/build_llvmlite.sh
@@ -21,9 +21,11 @@ conda activate $envname
 # Install llvmdev

 if [[ $(uname -m) == "aarch64" ]] ; then
-    conda install -y numba/label/manylinux_2_28::llvmdev --no-deps
+    #conda install -y numba/label/manylinux_2_28::llvmdev --no-deps
+    conda install -y -c file://$outputdir llvmdev --no-deps
 elif [[ $(uname -m) == "x86_64" ]] ; then
-    conda install -y numba/label/manylinux_2_17::llvmdev --no-deps
+    #conda install -y numba/label/manylinux_2_17::llvmdev --no-deps
+    conda install -y -c file://$outputdir llvmdev --no-deps
 else
     echo "Error: Unsupported architecture: $(uname -m)"
     exit 1
```

And then I ran:

```
./buildscripts/manylinux/docker_run_x64.sh build_llvmlite.sh cp313-cp313
```

To build and test llvmlite against a locally build llvmdev.

All tests passed! 🎉 

This concludes my functional spot test and this PR now LGTM!
